### ISSUE TITLE
Message From new phone number will now show in sms list

### ISSUE THREAD

**Issue by KruziikG** [moezbhatti/qksms#1200]
### DESCRIPTION
I noticed that sometimes QKSMS will not show new message in sms list and also there in no notification in status bar. Only if I chose Settings->Sync messages then these messages will show in QKSMS. I tried several different situations and finally know it only happens when you haven't have a conversion with the sender.
device: Huawei P9 Plus, running Android 8.0

### STEPS
1. Delete previous conversion with the sender if exist
2. Send a message to your phone using the number

### EXPECTED
The phone should produce a sound ringtone, and the new message should show in both notification bar and the QKSMS app.

### OBSERVATIONS
1. there is no ringtone, no notification , and you cannot find the message in QKSMS. But if you open another Messaging app, the message is displayed in that app.
2. return to QKSMS app, click ’Settings->Sync messages‘
3. finally the message will show in QKSMS

I will attach two logs,
the first one is captured in this situation: Screen is locked. I send three messages in this consequence :1. phone number A(new number） 2. phone number B(I have had conversion with this number privously）3.phone number C(another new number). Message 1 & 3 are now shown in QKSMS, but message 2 did. 
the second log is captured when phone is Waked up and QKSMS is in foreground. I Send a message using  phone number A. This time the message is also now shown in QKSMS.
[qksms_log1.txt](https://github.com/moezbhatti/qksms/files/2115523/qksms_log1.txt)
[qksms_log2.txt](https://github.com/moezbhatti/qksms/files/2115524/qksms_log2.txt)

this gif shows that when I finished send there 4 messages, I can only manual re-sync message.
![gif5](https://user-images.githubusercontent.com/11470789/41603798-4b2167ae-7410-11e8-8c72-94a8631ae841.gif)

**Comment by moezbhatti** [comment]
Fixed in 52c99faed130c3dcce899983e24cb6354da06e98

**Comment by moezbhatti** [comment]
If you delete the conversation and then receive a message from it again, does it show up in the app normally or do you need to sync it?

**Comment by KruziikG** [comment]
@moezbhatti I need to sync it

**Comment by moezbhatti** [comment]
@KruziikG QKSMS will only be able to stay in sync if it's set as the default SMS app

On Xiaomi devices when you open the default messaging app, it automatically assigns itself as the default SMS app, which means QKSMS won't receive your messages

I just tested this on my personal Xiaomi device using two different phones, and I can confirm that while it didn't work on 3.2.2, it works fine on [3.3.3](https://github.com/moezbhatti/qksms/releases/tag/v3.3.3)

**Comment by KruziikG** [comment]
@moezbhatti QKSMS is always the default SMS app. I think this only happen on some HUAWEI devices.

I added a log here:
![01](https://user-images.githubusercontent.com/11470789/41649704-c8cbfbd0-74ae-11e8-8da7-e6311386f889.png)

when add message to existing conversation, everything is OK
![02](https://user-images.githubusercontent.com/11470789/41649717-cf89c9a2-74ae-11e8-9cab-1cd0ebd96270.png)

but when start a new conversation, threadId  is 0, and this is an error:
(error message means : W/System.err: java.io.IOException: Do not send same content again.)
![03](https://user-images.githubusercontent.com/11470789/41649738-dca0f7f0-74ae-11e8-9610-622e938cd086.png)

Then I run this version using Android emulator. This app can start new conversation , and the difference is that the threadId for new conversation is not 0. like this
![04](https://user-images.githubusercontent.com/11470789/41650216-18fe737a-74b0-11e8-94c5-2e7ede51dcef.png)

**Comment by moezbhatti** [comment]
@KruziikG thank you for the detail! I'll make sure this gets fixed

**Comment by moezbhatti** [comment]
@KruziikG @gkomarnicki would you mind trying this APK and seeing if it resolves the issue for you?

[presentation-noAnalytics-universal-release.apk.zip](https://github.com/moezbhatti/qksms/files/2124070/presentation-noAnalytics-universal-release.apk.zip)

**Comment by gkomarnicki** [comment]
@moezbhatti The bug's still there. Android Messages and Pulse both got the SMS, while QKSMS didn't. QKSMS is the default SMS app

**Comment by moezbhatti** [comment]
@gkomarnicki hmmm, perhaps this one?

[presentation-noAnalytics-universal-release.apk.zip](https://github.com/moezbhatti/qksms/files/2124225/presentation-noAnalytics-universal-release.apk.zip)

**Comment by KruziikG** [comment]
@moezbhatti Yes, so weird. In fact, The android 8.0 of my P9 Plus is still a beta version. So I'm not sure whether it is a system bug.  Nova 2i  got android 8.0 beta 2 months before P9, and huawei has released Final version On 1 Jun. Maybe this problem is caused by huawei, and only happens in its EMUI 8.0 systems, who knows.

Another wired thing may be relatied. When using AOSP messaging app, All messages display as  'unknown sender'. But google messages works fine. 
![screenshot_20180621-222749](https://user-images.githubusercontent.com/11470789/41725575-b6250aae-75a2-11e8-9f47-b8203a70880f.jpg)

and today I tried to debug qksms, and find that not only the 'threadId' is 0, but also the conversation is null,
![default](https://user-images.githubusercontent.com/11470789/41727112-29ff7baa-75a6-11e8-9022-76480640d402.png)

-----------------------
and I tried both apks just now, this problen is still there. Here are the logs

[log06211.log](https://github.com/moezbhatti/qksms/files/2124167/log06211.log)
[log06212.log](https://github.com/moezbhatti/qksms/files/2124257/log06212.log)

**Comment by moezbhatti** [comment]
@KruziikG can you try replacing the `insertReceivedSms` method with this

``` kotlin
    override fun insertReceivedSms(subId: Int, address: String, body: String, sentTime: Long): Message {

        // Insert the message to the native content provider
        val uri = contentValuesOf(
                Pair(Telephony.Sms.ADDRESS, address),
                Pair(Telephony.Sms.BODY, body),
                Pair(Telephony.Sms.DATE_SENT, sentTime),
                Pair(Telephony.Sms.SUBSCRIPTION_ID, subId))
                .let { values -> context.contentResolver.insert(Telephony.Sms.Inbox.CONTENT_URI, values) }

        // Insert the message to Realm
        return Message().apply {
            this.address = address
            this.body = body
            this.dateSent = sentTime
            this.date = System.currentTimeMillis()
            this.subId = subId

            id = messageIds.newId()
            threadId = TelephonyCompat.getOrCreateThreadId(context, address)
            contentId = uri?.lastPathSegment?.toLong() ?: 0L
            boxId = Telephony.Sms.MESSAGE_TYPE_INBOX
            type = "sms"
            insertOrUpdate()
        }
    }
```

And can you let me know the value of `uri.toString()` and the value of `threadId`

**Comment by KruziikG** [comment]
I recall that my workmate has a huawei Nova young running EMUI 8.0 too, I can use his phone to test if his phone be affected. I know little about android app development, so I can only help to test. Just tell me if I can help.

**Comment by moezbhatti** [comment]
@KruziikG no worries, I assume that all of the affected devices will be affected in the same way. I'll send you code snippets and ask you to obtain certain values for me, this should help us resolve the issue by the end of the day :)

**Comment by moezbhatti** [comment]
Oh hm, that's interesting that it worked for you @gkomarnicki  but not @KruziikG 

@KruziikG let me know what the results are when you try that code snippet!

**Comment by KruziikG** [comment]
OK. And while syncing gradle files, I tried the second apk again, and I'm sure it not work for me. 
@gkomarnicki did you delete previous messages before test?

**Comment by KruziikG** [comment]
@moezbhatti both 'contentValuesOf' and 'values ' are not defined,don't know how to fix this.

![123123](https://user-images.githubusercontent.com/11470789/41731276-8cdf5a88-75b0-11e8-9332-26e5ff28f540.png)

**Comment by moezbhatti** [comment]
@KruziikG you'll need to update to the latest version of the dev branch first

**Comment by gkomarnicki** [comment]
@KruziikG I deleted most of the conversations from unknown numbers before testing, but not all of them.

![screenshot_20180622-060950](https://user-images.githubusercontent.com/20436870/41750590-5cf6d8f4-75ac-11e8-8a7d-5f219e6a4026.jpg)


In this thread where I received 2FA verification codes, there was no notification for the first two messages, but I got a notification for the third message after installing the most recent APK linked above.

**Comment by moezbhatti** [comment]
@KruziikG which version of android studio are you using?

**Comment by KruziikG** [comment]
@moezbhatti problem solved, I tried to imported 2 packages using alt+enter. 
But I think this project cannot be build on windows. I will check this later.

**Comment by moezbhatti** [comment]
@KruziikG thanks. let me know what you end up getting for those values

**Comment by KruziikG** [comment]
06-22 10:37:12.460 9566-9676/com.moez.QKSMS D/QKSMS: ------>content://sms/inbox//871
06-22 10:37:12.581 9566-9676/com.moez.QKSMS D/QKSMS: threadId-->522

now threadid below also has a value. but this message still not displayed.

**Comment by moezbhatti** [comment]
@KruziikG can you try this and share the logs

constructor
```kotlin
class MessageRepositoryImpl @Inject constructor(
        private val context: Context,
        private val cursorToConversation: CursorToConversation,
        private val messageIds: KeyManager,
        private val imageRepository: ImageRepository,
        private val prefs: Preferences) : MessageRepository {
```

insertReceivedSms
```kotlin
    override fun insertReceivedSms(subId: Int, address: String, body: String, sentTime: Long): Message {

        // Insert the message to the native content provider
        val uri = contentValuesOf(
                Pair(Telephony.Sms.ADDRESS, address),
                Pair(Telephony.Sms.BODY, body),
                Pair(Telephony.Sms.DATE_SENT, sentTime),
                Pair(Telephony.Sms.SUBSCRIPTION_ID, subId))
                .let { values -> context.contentResolver.insert(Telephony.Sms.Inbox.CONTENT_URI, values) }

        val threadId = TelephonyCompat.getOrCreateThreadId(context, address)

        val conversation1 = cursorToConversation.getConversationCursor(threadId)
                ?.map(cursorToConversation::map)
                ?.firstOrNull()

        val conversation2 = cursorToConversation.getConversationsCursor()
                ?.map(cursorToConversation::map)
                ?.firstOrNull { it.id == threadId }

        Timber.d("threadId: $threadId")
        Timber.d("conversation found via method 1: ${conversation1 != null}")
        Timber.d("conversation found via method 2: ${conversation2 != null}")

        // Insert the message to Realm
        return Message().apply {
            this.threadId = threadId
            this.address = address
            this.body = body
            this.dateSent = sentTime
            this.date = System.currentTimeMillis()
            this.subId = subId

            id = messageIds.newId()
            contentId = uri?.lastPathSegment?.toLong() ?: 0L
            boxId = Telephony.Sms.MESSAGE_TYPE_INBOX
            type = "sms"
            insertOrUpdate()
        }
    }
```

**Comment by KruziikG** [comment]
OK, here is the result
06-22 11:15:47.161 12557-12646/com.moez.QKSMS D/MessageRepositoryImpl: threadId: 522
06-22 11:15:47.162 12557-12646/com.moez.QKSMS D/MessageRepositoryImpl: conversation found via method 1: false
    conversation found via method 2: true

-----------------------------------------
and i found something before,in this function, the paremeter is valid, but it return null. And this issue is quite same as the comment said.
![screenshot from 2018-06-22 11-03-23](https://user-images.githubusercontent.com/11470789/41755932-4107b3ce-760c-11e8-926c-59b18d4093ea.png)

**Comment by KruziikG** [comment]
It Work! Great!
But.....Let me come up with a new issue about display, hahaha

**Comment by fazlerabbi37** [comment]
Hi there, I can see here that this issue has been closed with version 3.3.4 but I didn't find any update in F-Droid. When I chceked the vsersion info in F-Droid wiki, it say's they [couldn't build](https://f-droid.org/wiki/page/com.moez.QKSMS#Versions) the vsersion 3.3.4. Can you please check the issue?

**Comment by moezbhatti** [comment]
@fazlerabbi37 it gives the following reason, `Reason: uses com.android.tools.build:gradle:3.2.0-alpha18`

In 3.3.6 I've updated to `3.2.0-beta01`. Hopefully this helps, though I'm not sure what requirements they have for their automated builds
